WIP utils library updates (#431)
matter-labs/zksync-web-era-docs: docs/api/js/utils.md

# Utilities

The [utilities library](https://github.com/matter-labs/zksync-2-dev/blob/94701bd2fbc590f733346934cfbccae08fc62f1a/sdk/zksync-web3.js/src/utils.ts) contains essential utilities for building on zkSync Era.

:::info
- This document describes common functions and constants you may need.
- Functions used internally are not necessarily described.
- Check the code for the full list.
:::

## Use the library

Access the library by importing it into your scripts.

```typescript
import { utils } from "zksync-web3";
```

> ## Constants

Reviewer: ```

export const ZKSYNC_MAIN_ABI = new utils.Interface(require('../../abi/IZkSync.json').abi);
export const CONTRACT_DEPLOYER = new utils.Interface(require('../../abi/ContractDeployer.json').abi);
export const L1_MESSENGER = new utils.Interface(require('../../abi/IL1Messenger.json').abi);
export const IERC20 = new utils.Interface(require('../../abi/IERC20.json').abi);
export const IERC1271 = new utils.Interface(require('../../abi/IERC1271.json').abi);
export const L1_BRIDGE_ABI = new utils.Interface(require('../../abi/IL1Bridge.json').abi);
export const L2_BRIDGE_ABI = new utils.Interface(require('../../abi/IL2Bridge.json').abi);

export const BOOTLOADER_FORMAL_ADDRESS = '0x0000000000000000000000000000000000008001';
export const CONTRACT_DEPLOYER_ADDRESS = '0x0000000000000000000000000000000000008006';
export const L1_MESSENGER_ADDRESS = '0x0000000000000000000000000000000000008008';
export const L2_ETH_TOKEN_ADDRESS = '0x000000000000000000000000000000000000800a';

export const L1_TO_L2_ALIAS_OFFSET = '0x1111000000000000000000000000000000001111';

export const EIP1271_MAGIC_VALUE = '0x1626ba7e';

export const EIP712_TX_TYPE = 0x71;
export const PRIORITY_OPERATION_L2_TX_TYPE = 0xff;

export const MAX_BYTECODE_LEN_BYTES = ((1 << 16) - 1) * 32;

```

all constants from the list above also should be described here

```

### Useful addresses

#### ETH token layer 1

```typescript
export const ETH_ADDRESS = '0x0000000000000000000000000000000000000000';
```

#### ETH token alias on ZkSync Era

```typescript
export const L2_ETH_TOKEN_ADDRESS = '0x000000000000000000000000000000000000800a';
```

#### Bootloader

```ts
export const BOOTLOADER_FORMAL_ADDRESS = '0x0000000000000000000000000000000000008001';
```

#### Contract deployer

```ts
export const CONTRACT_DEPLOYER_ADDRESS = '0x0000000000000000000000000000000000008006';
```

#### L1 messenger

```ts
export const L1_MESSENGER_ADDRESS = '0x0000000000000000000000000000000000008008';
```

### Gas

#### `DEFAULT_GAS_PER_PUBDATA_LIMIT`

- Use a large amount of gas for signing on layer 2.
- The amount ensures any reasonable limit is accepted. 

:::info
- The operator is NOT required to use the actual value and can use any value up to that signed by the user.
:::

```typescript
export const DEFAULT_GAS_PER_PUBDATA_LIMIT = 50000;
```

#### `REQUIRED_L1_TO_L2_GAS_PER_PUBDATA_LIMIT`

Any amount is possible for L1->L2 transactions because the cost per gas is adjusted.

```ts
export const REQUIRED_L1_TO_L2_GAS_PER_PUBDATA_LIMIT = 800;

> ```

Reviewer: for some reasons there is no mentrion about these functions:
```
export function isETH(token: Address) {
    return token.toLowerCase() == ETH_ADDRESS || token.toLowerCase() == L2_ETH_TOKEN_ADDRESS;
}

export function sleep(millis: number) {
    return new Promise((resolve) => setTimeout(resolve, millis));
}

export function layer1TxDefaults() {
    return {
        queueType: PriorityQueueType.Deque,
        opTree: PriorityOpTree.Full
    };
}
```

Author: i will add these ->  i thought they were maybe uninteresting to users ...

## Functions

### `applyL1ToL2Alias`

Converts the address that submitted a transaction to the inbox on L1 to the `msg.sender` viewed on L2.

See also [`undol1tol2alias`](#undol1tol2alias).

```ts

> export function applyL1ToL2Alias(address: string): string {
>     return ethers.utils.hexlify(ethers.BigNumber.from(address).add(L1_TO_L2_ALIAS_OFFSET));
> }

Reviewer: ```suggestion
export function applyL1ToL2Alias(address: string): string {
    return ethers.utils.hexlify(ethers.BigNumber.from(address).add(L1_TO_L2_ALIAS_OFFSET).mod(ADDRESS_MODULO));
}
```

```

### `checkBaseCost`

Checks the base cost is within prescribed limits.

```ts
export async function checkBaseCost(baseCost: ethers.BigNumber, value: ethers.BigNumberish | Promise<ethers.BigNumberish>) {
    if (baseCost.gt(await value)) {
        throw new Error(
            `The base cost of performing the priority operation is higher than the provided value parameter ` +
                `for the transaction: baseCost: ${baseCost}, provided value: ${value}`
        );
    }
}
```

### `create2Address`

Generates a future-proof contract address using salt plus bytecode which allows determination of an address before deployment. 

:::warning
- The zkSync Era implementation is slightly different from Ethereum.
:::

#### Inputs

- `sender`: sender's address.
- `bytecodeHash`: output from zkSolc.
- `salt`: randomization element.

> - `input`: ABI encoded constructor arguments.

Reviewer: please add this section for other functions too, or do not sure is anywhere so it would be unified

Author: OK, I have done this @AnastasiiaVashchuk  ... for any inputs I had a question, I have put ?? when I'm finished going through your comments, I'll ask you to check these again if I may? Thank you

    } else {
        return await isEIP1271SignatureCorrect(provider, address, msgHash, signature);
    }
}
```

### `isTypedDataSignatureCorrect`

Returns true if account abstraction EIP712 signature is correct.

```ts
export async function isTypedDataSignatureCorrect(
    provider: Provider,
    address: string,
    domain: TypedDataDomain,
    types: Record<string, Array<TypedDataField>>,
    value: Record<string, any>,
    signature: SignatureLike
): Promise<boolean> {
    const msgHash = ethers.utils._TypedDataEncoder.hash(domain, types, value);
    return await isSignatureCorrect(provider, address, msgHash, signature);
}
```

### `parseTransaction`

Common parsing transaction function used by internal teams. 

Please see the [utilities library definition](https://github.com/matter-labs/zksync-2-dev/blob/94701bd2fbc590f733346934cfbccae08fc62f1a/sdk/zksync-web3.js/src/utils.ts) for more info.

### `serialize`

Common serialize function used by internal teams. 

Please see the [utilities library definition](https://github.com/matter-labs/zksync-2-dev/blob/94701bd2fbc590f733346934cfbccae08fc62f1a/sdk/zksync-web3.js/src/utils.ts) for more info.

> ### `serializeTransaction`

Reviewer: this is ethers function`node_modules/@ethersproject/transactions/src.ts/index.ts`

Reviewer: we do not need to mention it here

Author: Removed serializeTransaction function

Common serialize transaction function used by internal teams. 

Please see the [utilities library definition](https://github.com/matter-labs/zksync-2-dev/blob/94701bd2fbc590f733346934cfbccae08fc62f1a/sdk/zksync-web3.js/src/utils.ts) for more info.

### `undoL1ToL2Alias`

Converts the `msg.sender` viewed on L2 to the address that submitted a transaction to the inbox on L1.

See also [`applyl1tol2alias`](#applyl1tol2alias).

> ```ts
> export function undoL1ToL2Alias(address: string): string {
>     return ethers.utils.hexlify(ethers.BigNumber.from(address).sub(L1_TO_L2_ALIAS_OFFSET));
> }

Reviewer: up-to-date function version:
```suggestion
```ts
export function undoL1ToL2Alias(address: string): string {
    let result = ethers.BigNumber.from(address).sub(L1_TO_L2_ALIAS_OFFSET);
    if (result.lt(BigNumber.from(0))) {
        result = result.add(ADDRESS_MODULO);
    }

    return ethers.utils.hexlify(result);
}
```

```ts
export const EIP1271_MAGIC_VALUE = '0x1626ba7e';
```

#### EIP712 transaction type

Constant representing an EIP712 transaction type.

```ts
export const EIP712_TX_TYPE = 0x71;
```

#### Priority op transaction on L2

Constant representing a priority transaction operation on L2.

```ts
export const PRIORITY_OPERATION_L2_TX_TYPE = 0xff;
```

#### Max bytecode length

Used for ensuring bytecode length is not over the maximum allowed.

```ts
export const MAX_BYTECODE_LEN_BYTES = ((1 << 16) - 1) * 32;
```

### Useful addresses

#### ETH token layer 1

```typescript
export const ETH_ADDRESS = '0x0000000000000000000000000000000000000000';
```

#### ETH token alias on ZkSync Era

```typescript
export const L2_ETH_TOKEN_ADDRESS = '0x000000000000000000000000000000000000800a';
```

#### Bootloader

```ts
export const BOOTLOADER_FORMAL_ADDRESS = '0x0000000000000000000000000000000000008001';
```

#### Contract deployer

```ts
export const CONTRACT_DEPLOYER_ADDRESS = '0x0000000000000000000000000000000000008006';
```

#### L1 messenger

```ts
export const L1_MESSENGER_ADDRESS = '0x0000000000000000000000000000000000008008';
```

### Gas

#### `DEFAULT_GAS_PER_PUBDATA_LIMIT`

> - Use a large amount of gas for signing on layer 2.

Reviewer 2: ```suggestion
- Use a large amount of gas per pubdata for signing on layer 2.
```

- The amount ensures any reasonable limit is accepted. 

:::info
- The operator is NOT required to use the actual value and can use any value up to that signed by the user.
:::

```typescript
export const DEFAULT_GAS_PER_PUBDATA_LIMIT = 50000;
```

#### `REQUIRED_L1_TO_L2_GAS_PER_PUBDATA_LIMIT`

> Any amount is possible for L1->L2 transactions because the cost per gas is adjusted.

Reviewer 2: The current required gas per pubdata for L1->L2 transactions

Author: added

> The current required gas per pubdata for L1->L2 transactions. Any amount is possible for L1->L2 transactions because the cost per gas is adjusted.

Reviewer 2: Since this gas per pubdata is required, not any amount is possible : ) 

```suggestion
The current required gas per pubdata for L1->L2 transactions. 
```

- `provider`: ethers `Provider` object.

#### Outputs

An ABI-encoded array of:

- `nameBytes`: `bytes` object representation of token name.
- `symbolBytes`: `bytes` object representation of token symbol.
- `decimalBytes`: `bytes` object representation of token decimal representation.

```ts
async function getERC20GettersData(l1TokenAddress: string, provider: ethers.providers.Provider): Promise<string> {
    const token = IERC20MetadataFactory.connect(l1TokenAddress, provider);

    const name = await token.name();
    const symbol = await token.symbol();
    const decimals = await token.decimals();

    const coder = new AbiCoder();

    const nameBytes = coder.encode(['string'], [name]);
    const symbolBytes = coder.encode(['string'], [symbol]);
    const decimalsBytes = coder.encode(['uint256'], [decimals]);

    return coder.encode(['bytes', 'bytes', 'bytes'], [nameBytes, symbolBytes, decimalsBytes]);
}
```

### `getL2HashFromPriorityOp`

Returns the hash of the L2 priority operation from a given transaction receipt and L2 address.

#### Inputs

- `txReceipt`: receipt of the L1 transaction as ethers `TransactionReceipt` object.

> - `zkSyncAddress`: address of sender/transaction ?? on L2 as `Address` object.

Reviewer 2: this one is the address of the zkSync main contract on L1

Author: done

```ts
export function getL2HashFromPriorityOp(
    txReceipt: ethers.providers.TransactionReceipt,
    zkSyncAddress: Address
): string {
    let txHash: string = null;
    for (const log of txReceipt.logs) {
        if (log.address.toLowerCase() != zkSyncAddress.toLowerCase()) {
            continue;
        }

        try {
            const priorityQueueLog = ZKSYNC_MAIN_ABI.parseLog(log);
            if (priorityQueueLog && priorityQueueLog.args.txHash != null) {
                txHash = priorityQueueLog.args.txHash;
            }
        } catch {}
    }
    if (!txHash) {
        throw new Error('Failed to parse tx logs');
    }

    return txHash;
}
```

### `getHashedL2ToL1Msg`

Returns a keccak encoded message with a given sender address and block number from the L1 messenger contract.

#### Inputs

> - `sender`: sender address on L1 ?? as `Address` object.

Reviewer 2: ```suggestion
- `sender`: the sender of the message on L2 as `Address` string.
```

- `msg`: encoded message as `BytesLike` object.
- `txNumberInBlock`: index of the transaction in the block as number.

```ts
export function getHashedL2ToL1Msg(sender: Address, msg: BytesLike, txNumberInBlock: number) {
    const encodedMsg = new Uint8Array([
        0, // l2ShardId
        1, // isService
        ...ethers.utils.zeroPad(ethers.utils.hexlify(txNumberInBlock), 2),
        ...ethers.utils.arrayify(L1_MESSENGER_ADDRESS),
        ...ethers.utils.zeroPad(sender, 32),
        ...ethers.utils.arrayify(ethers.utils.keccak256(msg))
    ]);

    return ethers.utils.keccak256(encodedMsg);
}
```

### `getSignature`

Returns the signature of any transaction as a `Uint8Array`.

#### Inputs

> - `transaction`: transaction address ?? as ??.

Reviewer 2: Let's just not document this method, it is needed only internally. It is not a good thing that it is exposed to the external users, we should  remove it from the public API.

Author: It is called from elsewhere in the code. Any interested user will end up going to git to look it up where it is published, ie. it is already public. Are you sure?

Author: removed

    }

    const hashStr = ethers.utils.sha256(bytecodeAsArray);
    const hash = ethers.utils.arrayify(hashStr);

    // Note that the length of the bytecode
    // should be provided in 32-byte words.
    const bytecodeLengthInWords = bytecodeAsArray.length / 32;
    if (bytecodeLengthInWords % 2 == 0) {
        throw new Error('Bytecode length in 32-byte words must be odd');
    }

    const bytecodeLength = ethers.utils.arrayify(bytecodeLengthInWords);

    // The bytecode should always take the first 2 bytes of the bytecode hash,
    // so we pad it from the left in case the length is smaller than 2 bytes.
    const bytecodeLengthPadded = ethers.utils.zeroPad(bytecodeLength, 2);

    const codeHashVersion = new Uint8Array([1, 0]);
    hash.set(codeHashVersion, 0);
    hash.set(bytecodeLengthPadded, 2);

    return hash;
}
```

### `isECDSASignatureCorrect`

Like similar functionality in `ethers.js` but with added try/catch facility. The function returns true if the validation process succeeds.

Called from [`isSignatureCorrect`](#isSignatureCorrect) for non-contract account addresses. 

#### Inputs

> - `address`: sender address ?? as string.

Reviewer 2: the address which should've signed the `msgHash`

Author: done

Called from [`isMessageSignatureCorrect`](#ismessagesignaturecorrect) and [`isTypedDataSignatureCorrect`](#istypeddatasignaturecorrect). Returns true if account abstraction EIP712 signature is correct.

#### Inputs

- `provider`: `Provider` object.
- `address`: sender address as string.
- `msgHash`: hash of the message as string. 
- `signature`: ethers signature as `SignatureLike` object.

```ts
async function isSignatureCorrect(
    provider: Provider,
    address: string,
    msgHash: string,
    signature: SignatureLike
): Promise<boolean> {
    let isContractAccount = false;

    const code = await provider.getCode(address);
    isContractAccount = ethers.utils.arrayify(code).length != 0;

    if (!isContractAccount) {
        return isECDSASignatureCorrect(address, msgHash, signature);
    } else {
        return await isEIP1271SignatureCorrect(provider, address, msgHash, signature);
    }
}
```

### `isTypedDataSignatureCorrect`

Returns true if account abstraction EIP712 signature is correct.

#### Inputs

- `provider`: Provider object.
- `address`: sender address as string.
- `domain`: `TypedDataDomain` object from `@ethersproject/abstract-signer`.

> - `types`: map of records pointing to array of data as `Record<string, Array<TypedDataField>>` ??. 

Reviewer 2: map of recording pointing from the field's name to its type

Author: done

:::info
- This document describes common functions and constants you may need.
- Functions used internally are not necessarily described.
- Check the code for the full list.
:::

## Use the library

Access the library by importing it into your scripts.

```typescript
import { utils } from "zksync-web3";
```

## Constants

### Interfaces

#### zkSync Era main contract

```typescript
export const ZKSYNC_MAIN_ABI = new utils.Interface(require('../../abi/IZkSync.json').abi);
```

#### IERC20

For interacting with native tokens.

```typescript
export const IERC20 = new utils.Interface(require('../../abi/IERC20.json').abi);
```

#### IERC1271

```ts
export const IERC1271 = new utils.Interface(require('../../abi/IERC1271.json').abi);
```

#### Contract deployer

Used for deploying smart contracts.

```ts
export const CONTRACT_DEPLOYER = new utils.Interface(require('../../abi/ContractDeployer.json').abi);
```

#### L1 messenger

Used for sending messages from zkSync Era to Ethereum.

```ts
export const L1_MESSENGER = new utils.Interface(require('../../abi/IL1Messenger.json').abi);
```

#### L1 and L2 bridges

> Bridge interface abis for L1 and L2.

Reviewer 2: ```suggestion
Bridge interface ABIs for L1 and L2.
```

```ts
export const L1_BRIDGE_ABI = new utils.Interface(require('../../abi/IL1Bridge.json').abi);
export const L2_BRIDGE_ABI = new utils.Interface(require('../../abi/IL2Bridge.json').abi);
```

#### L1 to L2 alias offset

> Used for applying and undoing aliases on addresses from L1 to L2.

Reviewer 2: ```suggestion
Used for applying and undoing aliases on addresses during bridging from L1 to L2.
```

```ts
export const EIP712_TX_TYPE = 0x71;
```

#### Priority op transaction on L2

Constant representing a priority transaction operation on L2.

```ts
export const PRIORITY_OPERATION_L2_TX_TYPE = 0xff;
```

#### Max bytecode length

Used for ensuring bytecode length is not over the maximum allowed.

```ts
export const MAX_BYTECODE_LEN_BYTES = ((1 << 16) - 1) * 32;
```

### Useful addresses

#### ETH token layer 1

```typescript
export const ETH_ADDRESS = '0x0000000000000000000000000000000000000000';
```

#### ETH token alias on ZkSync Era

```typescript
export const L2_ETH_TOKEN_ADDRESS = '0x000000000000000000000000000000000000800a';
```

#### Bootloader

```ts
export const BOOTLOADER_FORMAL_ADDRESS = '0x0000000000000000000000000000000000008001';
```

#### Contract deployer

```ts
export const CONTRACT_DEPLOYER_ADDRESS = '0x0000000000000000000000000000000000008006';
```

#### L1 messenger

```ts
export const L1_MESSENGER_ADDRESS = '0x0000000000000000000000000000000000008008';
```

### Gas

#### `DEFAULT_GAS_PER_PUBDATA_LIMIT`

- Use a large amount of gas for signing on layer 2.
- The amount ensures any reasonable limit is accepted. 

:::info

> - The operator is NOT required to use the actual value and can use any value up to that signed by the user.

Reviewer 2: we want to change this at some point in time in the future

```suggestion
- As of now, the operator is NOT required to use the actual value and can use any value up to that signed by the user. This will be changed in the future.
```

Author: That's ok, we tell them then, not now.

:::

```typescript
export const DEFAULT_GAS_PER_PUBDATA_LIMIT = 50000;
```

#### `REQUIRED_L1_TO_L2_GAS_PER_PUBDATA_LIMIT`

Any amount is possible for L1->L2 transactions because the cost per gas is adjusted.

```ts
export const REQUIRED_L1_TO_L2_GAS_PER_PUBDATA_LIMIT = 800;
```

## Functions

### `applyL1ToL2Alias`

Converts the address that submitted a transaction to the inbox on L1 to the `msg.sender` viewed on L2.

#### Inputs

- `address`: sender address as string.

#### Outputs

> ?? not sure how to describe what's coming out here

Reviewer 2: "The `msg.sender` of the L1->L2 transaction is the `address` is a contract that initiated it."

As a remark:
During normal transaction if contract A calls contract B, the `msg.sender` of the call will be A.
During L1->L2 communication if an EOA X calls contract B, the `msg.sender` of the call will be X (the same logic as before).
During L1->L2 communication if a *contract* A calls contract B, the `msg.sender` of the call will be `applyL1ToL2Alias(A)` (the same logic as before).

Author: added

```ts
export function applyL1ToL2Alias(address: string): string {
    return ethers.utils.hexlify(ethers.BigNumber.from(address).add(L1_TO_L2_ALIAS_OFFSET).mod(ADDRESS_MODULO));
}
```

See also [`undol1tol2alias`](#undol1tol2alias).

> ### `checkBaseCost`

Reviewer 2: Unless there is a strict requiremet to document everything I would rather not document this method as it is rather low level and not needed for users

Author: My vote is leave it in, for a number of reasons. 
1. Even though users may never use it, they will see it mentioned here and there and perhaps wonder why it isn't documented, maybe go to git instead to look it up if they're interested. Our docs then look incomplete.
2. Our docs should be helpful to us too, new devs onboarding for example. Leaving info out again just makes them have to click around.
3. Completeness.

Reviewer 2: @kailash-manasarovar but if we document every internal detail, then every change to the SDK (even its internal parts which should never be used by anyone) will be considered a breaking change. Then we should clearly document under each such internal method that it is internal and may be changed at any moment

Author: good point, removed, the user can look these ones up at their leisure

Checks the base cost is within prescribed limits.

#### Inputs

> - `baseCost`: base cost ?? as `BigNumber` object.

Reviewer 2: You can call it "the minimal price of the L1->L2 transaction in ETH".

Author: done

> - `value`: value ?? as `BigNumberIsh` object.

Reviewer 2: The value to be provided with the L1 transaction

Author: ok

                `for the transaction: baseCost: ${baseCost}, provided value: ${value}`
        );
    }
}
```

### `create2Address`

Generates a future-proof contract address using salt plus bytecode which allows determination of an address before deployment. 

:::warning
- The zkSync Era implementation is slightly different from Ethereum.
:::

#### Inputs

- `sender`: sender address as string.
- `bytecodeHash`: output from zkSolc as `BytesLike` object.
- `salt`: randomization element as `BytesLike` object.
- `input`: ABI encoded constructor arguments as `BytesLike` object.

#### Outputs

- Returns an `Address` object.

```ts
export function create2Address(sender: Address, bytecodeHash: BytesLike, salt: BytesLike, input: BytesLike) {
    const prefix = ethers.utils.keccak256(ethers.utils.toUtf8Bytes('zksyncCreate2'));
    const inputHash = ethers.utils.keccak256(input);
    const addressBytes = ethers.utils
        .keccak256(ethers.utils.concat([prefix, ethers.utils.zeroPad(sender, 32), salt, bytecodeHash, inputHash]))
        .slice(26);
    return ethers.utils.getAddress(addressBytes);
}
```

:::tip

> The `prefix` in the `kekkak256` calculation is equal to `keccak256("zksyncCreate")`.

Reviewer 2: ```suggestion
The `prefix` is equal to `keccak256("zksyncCreate")`.
```

:::

### `createAddress`

Generates a contract address from deployer's account and nonce.

#### Inputs

- `sender`: sender address as string.
- `senderNonce`: sender nonce as `BigNumberish` object.

#### Outputs

- Returns an `Address` object.

```ts
export function createAddress(sender: Address, senderNonce: BigNumberish) {
    const prefix = ethers.utils.keccak256(ethers.utils.toUtf8Bytes('zksyncCreate'));
    const addressBytes = ethers.utils
        .keccak256(
            ethers.utils.concat([
                prefix,
                ethers.utils.zeroPad(sender, 32),
                ethers.utils.zeroPad(ethers.utils.hexlify(senderNonce), 32)
            ])
        )
        .slice(26);

    return ethers.utils.getAddress(addressBytes);
}
```

### `eip712TxHash`

> Returns signed keccak256 representation of a transaction with signature.

Reviewer 2: Returns the hash of an EIP712 transaction

Author: done

Author: ok, I updated the inputs too, please check

Author: @StanislavBreadless just this one to resolve i think

#### Inputs

> - `transaction`: given by id ?? as ??

Reviewer 2: the serialized EIP712 transaction object

> - `ethSignature?`: the transaction signer signature ?? as `EthereumSignature` object (optional).

Reviewer 2: optional ECDSA signature of the transaction

Author: done

```ts
function eip712TxHash(transaction: any, ethSignature?: EthereumSignature) {
    const signedDigest = EIP712Signer.getSignedDigest(transaction);
    const hashedSignature = ethers.utils.keccak256(getSignature(transaction, ethSignature));

    return ethers.utils.keccak256(ethers.utils.hexConcat([signedDigest, hashedSignature]));
}
```

### `estimateDefaultBridgeDepositL2Gas`

> Returns an estimation of gas required to execute a transaction from L1 to L2.

Reviewer 2: estimation of L2 gas needed to process the bridging of a token via the default ERC20 bridge (maybe link to the section about default bridges if there is one)

Author: done

#### Inputs

- `providerL1`: ethers `Provider` object.
- `providerL2`: zkSync `Provider` object.

> - `token`: token `Address` object.

Reviewer 2: it is a string, it is just type alias

Author: > updated, please check this

- `amount`: deposit amount as `BigNumberish`.

> - `to`: recipient `Address` object.

Reviewer 2: same here `Address` = `string`. It is just an alias for better readibility

Author: done

    const addressBytesLen = 40;
    const deployedContracts = receipt.logs
        .filter(
            (log) =>
                log.topics[0] == utils.id('ContractDeployed(address,bytes32,address)') &&
                log.address == CONTRACT_DEPLOYER_ADDRESS
        )
        // Take the last topic (deployed contract address as U256) and extract address from it (U160).
        .map((log) => {
            const sender = `0x${log.topics[1].slice(log.topics[1].length - addressBytesLen)}`;
            const bytesCodehash = log.topics[2];
            const address = `0x${log.topics[3].slice(log.topics[3].length - addressBytesLen)}`;
            return {
                sender: utils.getAddress(sender),
                bytecodeHash: bytesCodehash,
                deployedAddress: utils.getAddress(address)
            };
        });

    return deployedContracts;
}
```

### `getERC20BridgeCalldata`

Returns the calldata sent by an L1 ERC20 bridge to its L2 counterpart.

#### Inputs

- `l1TokenAddress`: token address on L1 as string.
- `l1Sender`: sender address on L1 as string.
- `l2Receiver`: recipient address on L2 as string.

> - `amount`: gas fee ?? as `BigNumberish` object.

Reviewer 2: the number of tokens to bridge

Author: added

- `provider`: ethers `Provider` object.

```ts
export async function getERC20BridgeCalldata(
    l1TokenAddress: string,
    l1Sender: string,
    l2Receiver: string,
    amount: BigNumberish,
    provider: ethers.providers.Provider
): Promise<string> {
    const gettersData = await getERC20GettersData(l1TokenAddress, provider);
    return L2_BRIDGE_ABI.encodeFunctionData('finalizeDeposit', [
        l1Sender,
        l2Receiver,
        l1TokenAddress,
        amount,
        gettersData
    ]);
}
```

### `getERC20GettersData`

Gets the data needed for initializing an L1 token counterpart on L2.

#### Inputs

- `l1TokenAddress`: token address on L1 as string.
- `provider`: ethers `Provider` object.

#### Outputs

> An array of:

Reviewer 2: ```suggestion
An ABI-encoded array of:
```

- `nameBytes`: `bytes` object representation of token name.
- `symbolBytes`: `bytes` object representation of token symbol.
- `decimalBytes`: `bytes` object representation of token decimal representation.

```ts
async function getERC20GettersData(l1TokenAddress: string, provider: ethers.providers.Provider): Promise<string> {
    const token = IERC20MetadataFactory.connect(l1TokenAddress, provider);

    const name = await token.name();
    const symbol = await token.symbol();
    const decimals = await token.decimals();

    const coder = new AbiCoder();

    const nameBytes = coder.encode(['string'], [name]);
    const symbolBytes = coder.encode(['string'], [symbol]);
    const decimalsBytes = coder.encode(['uint256'], [decimals]);

    return coder.encode(['bytes', 'bytes', 'bytes'], [nameBytes, symbolBytes, decimalsBytes]);
}
```

### `getL2HashFromPriorityOp`

Returns the hash of the L2 priority operation from a given transaction receipt and L2 address.

#### Inputs

> - `txReceipt`: ethers `TransactionReceipt` object.

Reviewer 2: it is the receipt of the L1 transaction (I think it is better to specify the meaning of the fields instead of just naming their types)

Author: sometimes it's obvious by the name itself, here it definitely needed more explanation, added

- `zkSyncAddress`: address of sender/transaction ?? on L2 as `Address` object.

```ts
export function getL2HashFromPriorityOp(
    txReceipt: ethers.providers.TransactionReceipt,
    zkSyncAddress: Address
): string {
    let txHash: string = null;
    for (const log of txReceipt.logs) {
        if (log.address.toLowerCase() != zkSyncAddress.toLowerCase()) {
            continue;
        }

        try {
            const priorityQueueLog = ZKSYNC_MAIN_ABI.parseLog(log);
            if (priorityQueueLog && priorityQueueLog.args.txHash != null) {
                txHash = priorityQueueLog.args.txHash;
            }
        } catch {}
    }
    if (!txHash) {
        throw new Error('Failed to parse tx logs');
    }

    return txHash;
}
```

### `getHashedL2ToL1Msg`

Returns a keccak encoded message with a given sender address and block number from the L1 messenger contract.

#### Inputs

- `sender`: sender address on L1 ?? as `Address` object.
- `msg`: encoded message as `BytesLike` object.

> - `txNumberInBlock`: block number containing message on L2 ?? as number.

Reviewer 2: it is the index of the transaction in block

Author: added

    isContractAccount = ethers.utils.arrayify(code).length != 0;

    if (!isContractAccount) {
        return isECDSASignatureCorrect(address, msgHash, signature);
    } else {
        return await isEIP1271SignatureCorrect(provider, address, msgHash, signature);
    }
}
```

### `isTypedDataSignatureCorrect`

Returns true if account abstraction EIP712 signature is correct.

#### Inputs

- `provider`: Provider object.
- `address`: sender address as string.
- `domain`: `TypedDataDomain` object from `@ethersproject/abstract-signer`.
- `types`: map of records pointing to array of data as `Record<string, Array<TypedDataField>>` ??. 
- `value`: a single `Record` value as `Record<string, any>`.
- `signature`: ethers signature as `SignatureLike` object.

```ts
export async function isTypedDataSignatureCorrect(
    provider: Provider,
    address: string,
    domain: TypedDataDomain,
    types: Record<string, Array<TypedDataField>>,
    value: Record<string, any>,
    signature: SignatureLike
): Promise<boolean> {
    const msgHash = ethers.utils._TypedDataEncoder.hash(domain, types, value);
    return await isSignatureCorrect(provider, address, msgHash, signature);
}
```

> ### `layer1TxDefaults`

Reviewer 2: don't document this one (it is outdated), please create a task to remove this method

Author: removed
